README.md
# nmt-wizard

`nmt-wizard` is a Docker-based task launcher and monitor on a variety of remote platforms (called *services*) such as SSH servers, Torque clusters, or EC2 instances. Each *service* is providing access to compute *resources*. The launcher is meant to be used with `nmt-wizard-docker` images, but without a strong dependency.

The project provides:

* a RESTful server that queues incoming requests in a Redis database;
* a client to the REST server providing a simple textual visualization interface;
* workers that launch and manage tasks on the requested service and updates their status.

Once launched, the tasks are sending progress and updated information to the launcher.

## Services configuration

Service configurations are provided by the system administrators. A service is declared and configured by a JSON file in the `config` directory. The REST server and worker automatically discover existing configuration files, provided that their filename ends with `.json`. A special `default.json` file defines parameters shared by all services.

The configuration file has the following structure:

```
{
    "name": "my-service",  // The short name the user will select.
    "description": "My service",  // Display name of the service.
    "module": "services.XXX",  // Name of the Python module managing the service.
    "variables": { // Non-Service specific configurations.
        "key1": [ "value1", "value2" ],
        ...
    },
    "docker": {
        "registries": {  // Docker registries: ECS, Docker Hub.
            "aws": {
                "type": "aws",
                "credentials": {
                    "AWS_ACCESS_KEY_ID": "XXXXX",
                    "AWS_SECRET_ACCESS_KEY": "XXXXX"
                },
                "uri": "XXXXX.dkr.ecr.eu-west-3.amazonaws.com",
                "region": "eu-west-3"
            },
            "dockerhub": {
                "type": "dockerhub",
                "uri": "",
                "default_for": [ "nmtwizard" ]
            },
            "mydockerprivate": {
                "type": "dockerprivate",
                "uri": "",
                "credentials": {
                    "password": "XXXXX",
                    "username": "XXXXX"
                },
                "default_for": [ "systran" ]
            }
        },
        "mount": [  // Volumes to mount when running the Docker image.
            "/home/devling/corpus:/root/corpus",
            "/home/devling/models:/root/models"
        ],
        "envvar": {  // Environment variables to set when running the Docker image.
            "key1": "value1",
            "specific": {
                "image": {
                    "key": "value"
                }
            }
        },
        "path": "" // if provided, will be used to select default path for docker command on remote service.
    },
    "skey1": "svalue1",  // Non-Service specific configurations.
    ...,
    "disabled": [01],  // Boolean field to disable/enable the service.
    "storages": {  // Storage configuration as described in single-training-docker.
    },
    "cpu_allocation": { // define default number of cpu allocated for the different tasks
          "gpu_task": 2,
          "preprocess_task": 4,
          "trans_task": 4,
          "buildvocab_task": 4,
          "release_task": 2
    },
    "callback_url": "http://LAUNCHER_URL",
    "callback_interval": 60,
    "ttl_policy" : [ // define the time to live policy for completed task in the database
        { 
            "pattern": {
                "status": "stopped",
                "message": "dependency_error"
            },
            "ttl": 10
        }
    ]
}
```

where `variables` is a list of possible options for the service. The structure of some of these options is specific to each service. Common options are described below.

Template files are provided in `config/templates` and can be used as a basis for configuring services.

**Note**:
* it is possible to add a field `"default_ms":true` to one storage definition. If no model storage parameter (`-ms`) is provided to the docker, this storage will be used by default.
* to avoid exposing credentials in the task logs, annotate with "[[private:xxxx]]" all the values that are sensitive, these values will be masked as "[[private]]" in training logs.
* specific environment variables can be passed to specific image (for instance credentials) - use "specific" dict for that.
* `ttl_policy` defines the time to leave (ttl) for stopped tasks in the redis database. Each pattern defines a config to match on the stopped task, if all conditions applies, the corresponding ttl is applied.

## Server configuration

* The REST server and workers are configured by `settings.ini`. The default configuration for all of the pools is stored in `default.json`. This file is read by the launcher when the launcher is started and stored in REDIS database together with a timestamp. Whenever this file is changed, the corresponding timestamp must be modified - and this modification will be detected by all workers that are then restarting.

## Using the launcher

### Worker

The first component to launch is the worker that should always be running. It handles:

* the launch of tasks
* the termination of tasks
* the update of active resources

```bash
cd server && python worker.py configfile.json
```

One worker has to be launched for every service pool. To monitor each worker and allow to be relaunched in case it stops (either because of an issue/exception in the code, or simply for updating its parameter) - the worker should be launcher using `runworker.py` as following:

```bash
python runworker.py service_name
```

`runworker.py` and `worker.py` are requesting following:
* a `configurations` directory in which the different configurations for the service will be copied
* a `logs` directory in which worker logs will be saved
* an active configuration file for the service in the current directory with name `service_name.json`. Not that a service name should be composed of letters, digits, - and _ characters only
* the settings.ini file
* the logger configuration file `logging.conf`

To administer multiple services: create a main directory with above-mentioned structure, and launch as many workers as pool to create.

### Server

The server has the following HTTP routes:

| METHOD    | ROUTE | PARAM | Description |
| ---       | --- | --- | --- |
| `GET`     | `status` | | Check service availability |
| `GET`     | `service/list` | | Returns available services |
| `GET`     | `service/describe/{service_id}` | | Returns user selectable options for a specified service |
| `GET`     | `service/check/{service_id}` | | Checks availability of a given service with provided user options |
| `POST`    | `task/launch` | |  Launches a task on a given service with provided user options
| `GET`     | `task/status/{task_id}` | |  Checks the status of a task |
| `GET`     | `task/list` | prefix | Returns the list of tasks in the database |
| `GET`     | `task/terminate/{task_id}` | |  Terminates a task, the process and/or instance associated with a task |
| `DELETE`  | `task/{task_id}` | | Delete a task from the database |
| `PUT`     | `task/beat/{task_id}` | | Provides a `beat` back to the launcher to notify the task activity and announce the next beat to expect |
| `GET`<br>`POST` | `task/file/{task_id}` | | Gets or set a file associated to a task |
| `POST` | `task/stat/{task_id}` | | post json statistics for a task |
| `GET`<br>`PATCH`<br>`POST` | `task/log/{task_id}` | | Gets, appends or sets the log associated to a task |
| `GET`     | `service/listconfig/{service}` | | List all configurations associated to a given service | 
| `GET`     | `service/selectconfig/{service}/{configname}` | | Select a named configuration associated to a given service | 
| `POST`     | `service/addconfig/{service}/{configname}` | `config` | Register a new named configuration for a given service | 
| `GET`     | `service/delconfig/{service}/{configname}` | | Delete a new named configuration for a given service | 
| `GET`     | `service/restart/{service}` | | Restart a service | 
| `GET`     | `service/disable/{service}/{resource}` | `message`| Disable a resource for a given service  | 
| `GET`     | `service/enable/{service}/{resource}` | | Enable a disabled resource for a given service  |


The server uses Flask. See the [Flask documentation](http://flask.pocoo.org/docs/0.12/deploying/) to deploy it for production. For development purpose, it can be run as follows (single thread):

```bash
export LAUNCHER_CONFIG=${PWD}/settings.ini
FLASK_APP=main.py flask run [--host=0.0.0.0]
```

Here are the available routes. Also see the next section

#### `GET /service/list`

Lists available services.

* **Arguments:** None
* **Input:** None
* **Output:** A dictionary of service name to description (JSON), their usage and capacity
* **Example:**

```
$ curl -X GET 'http://127.0.0.1:5000/service/list'
{
    "demogpu02": {
        "capacity": 1,
        "name": "OVH-hosted extra training server",
        "queued": 2,
        "usage": 1
    },
    "localhost": {
        "capacity": 2,
        "name": "test local environment",
        "queued": 0,
        "usage": 0
    },
    "ec2": {
        "capacity": 15,
        "name": "Instance on AWS EC2",
        "queued": 0,
        "usage": 7
    }
}
```

#### `GET /service/describe/<service_name>`

Returns possible options for a service as a [JSON Form](https://github.com/joshfire/jsonform). This can be used to easily implement a GUI to select options for the target service.

* **Arguments:**
  * `service_name`: the service name
* **Input:** None
* **Output:** A JSON form (or an empty dictionary if the service has no possible options).
* **Example:**

```
$ curl -X GET 'http://127.0.0.1:5000/service/describe/ec2'
{
  "launchTemplate": {
    "description": "The name of the EC2 launch template to use",
    "enum": [
      "SingleTrainingDev"
    ],
    "title": "EC2 Launch Template",
    "type": "string"
  }
}
```

#### `GET /service/check/<service_name>`

Checks if the service is available and can be used with the provided options. In case of success, it returns information about the service and the corresponding resource.

* **Arguments:**
  * `service_name`: the service name
* **Input:** The selected service options (see `service/describe/<service_name>`) (JSON)
* **Output:**
  * On invalid option, a HTTP 400 code with the error message (JSON)
  * On server error, a HTTP 500 code with the error message (JSON)
  * On success, an optional message with details about the service (JSON)
* **Example:**

```
$ curl -X GET http://127.0.0.1:5000/service/check/ec2
{
  "message": "missing launchTemplateName option",
}
$ curl -X GET -d '{"launchTemplateName": "InvalidLaunchTemplate"}' \
    -H "Content-Type: application/json" 'http://127.0.0.1:5000/service/check/ec2'
{
  "message": "An error occurred (InvalidLaunchTemplateId.NotFound) when calling the RunInstances operation: LaunchTemplate null not found"
}
$ curl -X GET -d '{"launchTemplateName": "SingleTrainingDev"}' \
    -H "Content-Type: application/json" 'http://127.0.0.1:5000/service/check/ec2'
{
  "message": ""
}
```

#### `POST /task/launch/<service_name>`

Launches a Docker-based task on the specified service. In case of success, it returns a task identifier that can be used to monitor the task using the `status` or `terminate` routes.

* **Arguments:**
  * `service_name`: the service name
* **Input:** the input is either a simple json body or a multi-part request with `content` field containing JSON task configuration. The other fields of the multi-part requests are binary files to be uploaded on the remote service at task-launch time.

The task configuration (JSON)

```
$ cat body.json
{
  "docker": {
    "registry": "dockerhub"
    "image": "opennmt/opennmt-lua",
    "tag": "latest",
    "command": [
      ...
    ]
  },
  "wait_after_launch": 2,
  "trainer_id": "OpenNMT",
  "options": {
    "launchTemplateName": "SingleTrainingDev"
  },
  "name": 'TaskName', // optional
  "iterations": 4,    // number of training iterations, default 1
  "priority": 100     // task priority
}
```

`docker.tag` and `wait_after_launch` are optional.
* **Output:**
  * On invalid task configuration, a HTTP 400 code with an error message (JSON)
  * On success, a task identifier (string)
* **Example:**

```
$ curl -X POST -d @invalid_body.json -H "Content-Type: application/json" \
    http://127.0.0.1:5000/task/launch/ec2
{
  "message": "missing trainer_id field"
}
$ curl -X POST -d @body.json -H "Content-Type: application/json" \
    'http://127.0.0.1:5000/task/launch/ec2'
"SSJS_enyy_HelloWorld_01_0f32d3f6b84ab91d4"
$ curl -X POST -d content=@body.json -F input.txt=@input.txt 'http://127.0.0.1:5000/task/launch/ec2'
"SSJS_xxyy_GreenCat_01_085a8:60c06412a2b74"
```

**Notes**: 

* the task identifier is structured, when possible, and contains the following 5 fields separated by `_`:
  * `TID` - trainer identifier provided through client application
  * `XXYY` - the language pair, found in the configuration file or from parent model
  * `NAME` - model name: generated randomly, or set manually, or inherited from parent
  * `NN` - number showing the number of iterations
  * `UUID` or `UUID:PRID`, unique identifier (possibly suffixed by initial of parent UUID)
* if a `iterations` value is passed to the launch service, several tasks will be created each one starting with previous generated one. The tasks are executed iteratively. It is also possible to use a non-yet generated model as a starting point to launch another task: in that case, the task will start only upon successful termination of the parent task.

#### `GET /task/list/<pattern>`

Lists available services.

* **Arguments:**
  * `pattern`: pattern for the tasks to match. See [KEYS pattern](https://redis.io/commands/keys) for syntax.
* **Input:** None
* **Output:** A list of tasks matching the pattern with minimal information (`task_id`, `queued_time`, `status`, `service`, `message`, `pid`)
* **Example:**

```
$ curl -X GET 'http://127.0.0.1:5000/task/list/jean_*'
[
  {
    "message": "completed", 
    "queued_time": "1519652594.957615", 
    "status": "stopped",
    "service": "ec2",
    "task_id": "jean_5af69495-3304-4118-bd6c-37d0e6"
  }, 
  {
    "message": "error", 
    "queued_time": "1519652097.672299", 
    "status": "stopped",
    "service": "mysshgpu", 
    "task_id": "jean_99b822bc-51ac-4049-ba39-980541"
  }
]
```

#### `DELETE /task/<pattern>`

Lists available services.

* **Arguments:**
  * `pattern`: pattern for the tasks to match - only stopped tasks will be deleted. See [KEYS pattern](https://redis.io/commands/keys) for syntax.
* **Input:** None
* **Output:** list of deleted tasks
* **Example:**

```
$ curl -X DELETE 'http://127.0.0.1:5000/task/jean_*'
[
  "jean_5af69495-3304-4118-bd6c-37d0e6",
  "jean_99b822bc-51ac-4049-ba39-980541"
]
```

#### `GET /task/status/<task_id>`

Returns the status of a task.

* **Arguments:**
  * `task_id`: the task ID returned by `/task/launch/<service_name>`
* **Input:** None
* **Output:**
  * On invalid `task_id`, a HTTP 404 code dictionary with an error message (JSON)
  * On success, a dictionary with the task status (JSON)
* **Example:**

```
curl -X GET http://127.0.0.1:5000/task/status/unknwon-task-id
{
  "message": "task unknwon-task-id unknown"
}
curl -X GET http://127.0.0.1:5000/task/status/130d4400-9aad-4654-b124-d258cbe4b1e3
{
  "allocated_time": "1519148201.9924579",
  "content": "{\"docker\": {\"command\": [], \"registry\": \"dockerhub\", \"image\": \"opennmt/opennmt-lua\", \"tag\": \"latest\"}, \"service\": \"ec2\", \"wait_after_launch\": 2, \"trainer_id\": \"OpenNMT\", \"options\": {\"launchTemplateName\": \"SingleTrainingDev\"}}", 
  "message": "unknown registry",
  "queued_time": "1519148144.483396",
  "resource": "SingleTrainingDev",
  "service": "ec2",
  "status": "stopped",
  "stopped_time": "1519148201.9977396",
  "ttl": null
}
```

(Here the task was quickly stopped due to an incorrect Docker registry.)

The main fields are:

* `status`: (timestamp for each status can be found in `<status>_time`)
  * `queued`,
  * `allocated`,
  * `running`,
  * `terminating`,
  * `stopped` (additional information can be found in `message` field);
* `service`: name of the service the task is running on;
* `resource`: name of the resource the task is using;
* `content`: the actual task definition;
* `update_time`: if the task is sending beat requests;
* `ttl` if a time to live was passed in the beat request.

#### `GET /task/terminate/<task_id>(?phase=status)`

Terminates a task. If the task is already stopped, it does nothing. Otherwise, it changes the status of the task to `terminating` (actual termination is asynchronous) and returns a success message.

* **Arguments:**
  * `task_id`: the task identifier returned by `/task/launch/<service_name>`
  * (optionnal) `phase`: indicate if the termination command is corresponding to an error or natural completion (`completed`)
* **Input**: None
* **Output**:
  * On invalid `task_id`, a HTTP 404 code with an error message (JSON)
  * On success, a HTTP 200 code with a message (JSON)

```
curl -X GET http://127.0.0.1:5000/task/terminate/130d4400-9aad-4654-b124-d258cbe4b1e3
{
  "message": "130d4400-9aad-4654-b124-d258cbe4b1e3 already stopped"
}
```

#### `DELETE /task/<task_id>`

Deletes a task. If the task is not stopped, it does nothing.

* **Arguments:**
  * `task_id`: the task identifier returned by `/task/launch/<service_name>`
* **Input**: None
* **Output**:
  * On invalid `task_id`, a HTTP 404 code with an error message (JSON)
  * On success, a HTTP 200 code with a message (JSON)

#### `PUT /task/beat/<task_id>(?duration=XXX&container_id=CID)`

Notifies a *beat* back to the launcher. Tasks should invoke this route wih a specific interval to notify that they are still alive and working. This makes it easier for the launcher to identify and handle dead tasks.

* **Arguments**
  * `task_id`: the task identifier returned by `/task/launch/<service_name>`
  * (optional) `duration`: if no beat is received for this task after this duration the task is assumed to be dead
  * (optional) `container_id`: the ID of the Docker container
* **Input:** None
* **Output:**
  * On invalid `duration`, a HTTP 400 code with an error message (JSON)
  * On invalid `task_id`, a HTTP 404 code with an error message (JSON)
  * On success, a HTTP 200 code

#### `POST /task/file/<task_id>/<filename>`

Registers a file for a task - typically used for log, or posting translation output using http storage.

* **Arguments**
  * `task_id`: the task identifier returned by `/task/launch/<service_name>`
  * `filename`: a filename
* **Input:** None
* **Output:**
  * On invalid `task_id`, a HTTP 404 code with an error message (JSON)
  * On success, a HTTP 200 code

#### `GET /task/file/<task_id>/<filename>`

Retrieves file attached to a task

* **Arguments**
  * `task_id`: the task identifier returned by `/task/launch/<service_name>`
  * `filename`: a filename
* **Input:** None
* **Output:**
  * On invalid `task_id`, a HTTP 404 code with an error message (JSON)
  * On missing files, a HTTP 404 code with an error message (JSON)
  * On success, the actual file

#### `GET/POST/PATCH /task/log/<task_id>`

Gets/Posts/Append log attached to a task. Logs are saved in a special `log:<task_id>` key in the redis table allowing for fast implementation of append operation.

### Launcher

The launcher is a simple client to the REST server. See:

```bash
python client/launcher.py -h
```

**Notes:**


CHANGELOG.md
## Unreleased
* (client/launcher) global pool (#53416)
* (client/launcher) add entity in resource list command (#53576)
* (client) add snw client commands to share and open models (#53629)
* (client/launcher) python 3 (#53329)
* (client/launcher) coding style (#53529)

## [v1.10.2](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.10.1) (2019-12-18)
### Fixes and improvements
* (client/launcher) add `-e/--entity-owner` option when training a model(#52926)
* (worker) fix deadlock CPU allocation(#53071)

## [v1.10.1](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.10.1) (2019-10-21)
### Fixes and improvements
* (client) fix error with option service list(#52908)

## [v1.10.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.10.0) (2019-10-08)
### Fixes and improvements
* (launcher/client) update to support evaluating local (multi-) reference files
* (launcher/client) filter the task list by status, service (#52807)
* (launcher/client) show the parent task in the task list (#52807)
* (launcher) add the possibility to create a folder in S3 storages
* (launcher) store the information relative to the release action when a model is released (#52793)
* (launcher) Flask server returns proper error message (with the code 500) after an uncatched error (#52742)
* (launcher) prevent user to use resource/upload in pn9-model-catalog and all shared storages (except for systran resource manager )
* (launcher) create cache to improve the search of model

## [v1.9.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.9.0) (2019-07-23)
### Fixes and improvements
* (launcher) update to use latest storage management
* (launcher) fix S3 storage with "assume_role" feature (#52401)
* (launcher) limit task log access to entity's users (#52079)
* (launcher) fix crash when updating with empty statistics (#52209)

## [v1.8.1](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.8.1) (2019-06-18)
### Fixes and improvements
* (launcher) allow user see all tasks launched in his entity (#52201)

## [v1.8.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.8.0) (2019-06-14)
### Fixes and improvements
* (launcher) fix scoring task when working with launcher as storage
* (launcher) check ability when deleting task
* (launcher) fix crash when  alloc_ressource and resource not found in task
* correct inappropriate chinese funny name
* (client) remove debug (== True) displaying when launching operation involving local file

## [v1.7.3](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.7.3) (2019-05-14)
### Fixes and improvements
* (worker) fix worker not auto restart when "admin:worker:" redis key expired

## [v1.7.2](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.7.2) (2019-04-05)
### Fixes and improvements
* (client) fix task launch command

## [v1.7.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.7.0) (2019-04-02)
### Fixes and improvements
* add generic DB cache with redis 
* fix display of task list if some tasks have corrupted information
* (worker) fix crash for resource calculation
* (worker) fix use of single quote in configuration file
* (launcher) enable /file and /task/file equivalent API for backward compatibility
* (launcher) normalize naming of tasks with parent task id (#51648)
* add score docker lang option

## [v1.6.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.6.0) (2018-02-11)

### New features
* add "task/stat" route for tracking tasks statistics

### Fixes and improvements
* (worker) fix passing envvar to specific image as defined in non default config.
* (launcher) fix `lr` for s3 for root path

## [v1.5.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.5.0) (2018-01-28)

### New features
* add "exec" command to launch any generic nmt-wizard docker task

### Fixes and improvements
* (client) avoid content corruption when `file` output string contains non utf-8 or non supported characters
* (client) improve task list to add model and image tag information
* (client) `-l DEBUG` now traces HTTP requests
* (worker) fix: ignore `envvar` field in config.docker
* (admin) global `default.json` now stored in REDIS db and updated by launcher
* (server) introduce Chinese funnyname and generated model contains name translation

## [v1.4.2](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.4.2) (2018-12-17)

### Fixes and improvements

* (worker) fix delete of running CPU tasks from `ls` when worker restart

## [v1.4.1](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.4.1) (2018-12-17)

### Fixes and improvements
* Add French and German funny names
* (worker) fix issue with CPU allocation making tasks fail when all CPU were allocated
* (worker) worker does not fail if configuration is invalid, pool is not disabled
* (worker) do not set cpu restriction on dedicated instances (like ec2)
* (worker) double check process termination after docker terminate
* (worker) patch log renew beat
* (launcher) `--as_release` is now implied for translations for image >= 1.80, except if `--notransasrelease` is specified

## [v1.4.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.4.0) (2018-12-14)

### Fixes and improvements
* (worker) refactor and improve ec2 service

## [v1.3.1](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.3.1) (2018-12-11)

* (client) backward compatibility for `lt`

## [v1.3.0](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.3.0) (2018-12-10)

* (worker) refactor cpu allocation - are now allocated like gpu
* (launcher) automatic translation tasks are now by default launched with `--as_release`

## [v1.2.5](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.2.5) (2018-11-28)

### Fixes and improvements
* (worker) fix retry on ssh connection
* (worker) more robust environment checking before launching task

## [v1.2.4](https://github.com/OpenNMT/nmt-wizard/releases/tag/v1.2.4) (2018-11-26)

### Fixes and improvements
* (launcher) fix crash on `task/status` without parameters


server/main.py
import os
import logging
import flask
import json

from six.moves import configparser

from nmtwizard import common, config, task
from nmtwizard.redis_database import RedisDatabase
from nmtwizard.helper import build_task_id, shallow_command_analysis, change_parent_task

ch = logging.StreamHandler()
ch.setLevel(logging.ERROR)
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
ch.setFormatter(formatter)

config.add_log_handler(ch)
common.add_log_handler(ch)

cfg = configparser.ConfigParser()
cfg.read('settings.ini')
MODE = os.getenv('LAUNCHER_MODE', 'Production')

redis_password = None
if cfg.has_option(MODE, 'redis_password'):
    redis_password = cfg.get(MODE, 'redis_password')

redis = RedisDatabase(cfg.get(MODE, 'redis_host'),
                      cfg.getint(MODE, 'redis_port'),
                      cfg.get(MODE, 'redis_db'),
                      redis_password)
services = config.load_services(cfg.get(MODE, 'config_dir'))


def _get_service(service):
    """Wrapper to fail on invalid service."""
    if service not in services:
        response = flask.jsonify(message="invalid service name: %s" % service)
        flask.abort(flask.make_response(response, 404))
    return services[service]

def _usagecapacity(service):
    """calculate the current usage of the service."""
    usage = 0
    capacity = 0
    for resource in service.list_resources():
        capacity += service.list_resources()[resource]
        usage += redis.llen("resource:%s:%s" % (service.name, resource))
    queued = redis.llen("queued:"+service.name)
    return usage, queued, capacity

app = flask.Flask(__name__)

def task_request(func):
    def func_wrapper(*args, **kwargs):
        if not task.exists(redis, args[0]):
            flask.abort(flask.make_response(flask.jsonify(message="task %s unknown" % args[0]), 404))
        return func(*args, **kwargs)
    return func_wrapper

@app.route("/service/list", methods=["GET"])
def list_services():
    res = {}
    for k in services:
        usage, queued, capacity = _usagecapacity(services[k])
        res[k] = { 'name':services[k].display_name,
                   'usage': usage, 'queued': queued, 'capacity': capacity }
    return flask.jsonify(res)

@app.route("/service/describe/<string:service>", methods=["GET"])
def describe(service):
    service_module = _get_service(service)
    return flask.jsonify(service_module.describe())

@app.route("/service/check/<string:service>", methods=["GET"])
def check(service):
    service_options = flask.request.get_json() if flask.request.is_json else None
    if service_options is None:
        service_options = {}
    service_module = _get_service(service)
    try:
        details = service_module.check(service_options)
    except ValueError as e:
        flask.abort(flask.make_response(flask.jsonify(message=str(e)), 400))
    except Exception as e:
        flask.abort(flask.make_response(flask.jsonify(message=str(e)), 500))
    else:
        return flask.jsonify(message=details)

@app.route("/task/launch/<string:service>", methods=["POST"])
def launch(service):
    content = None
    files = {}
    if flask.request.is_json:
        content = flask.request.get_json()
    else:
        content = flask.request.form.get('content')
        if content is not None:
            content = json.loads(content)
        for k in flask.request.files:
            files[k] = flask.request.files[k].read()
    if content is None:
        flask.abort(flask.make_response(flask.jsonify(message="missing content in request"), 400))
    service_module = _get_service(service)
    content["service"] = service

    task_type = '????'
    if "train" in content["docker"]["command"]: task_type = "train"
    elif "trans" in content["docker"]["command"]: task_type = "trans"
    elif "preprocess" in content["docker"]["command"]: task_type = "prepr"

    # Sanity check on content.
    if 'options' not in content or not isinstance(content['options'], dict):
        flask.abort(flask.make_response(flask.jsonify(message="invalid options field"), 400))
    if 'docker' not in content:
        flask.abort(flask.make_response(flask.jsonify(message="missing docker field"), 400))
    if ('image' not in content['docker']
        or 'registry' not in content['docker']
        or 'tag' not in content['docker']
        or 'command' not in content['docker']):
        flask.abort(flask.make_response(flask.jsonify(message="incomplete docker field"), 400))
    if content['docker']['registry'] == 'auto':
        repository = content['docker']['image']
        p = repository.find("/")
        if p == -1:
            flask.abort(flask.make_response(flask.jsonify(message="image should be repository/name"), 400))
        repository = repository[:p]
        registry = None
        for r in service_module._config['docker']['registries']:
            v = service_module._config['docker']['registries'][r]
            if "default_for" in v and repository in v['default_for']:
                registry = r
                break
        if registry is None:
            flask.abort(flask.make_response(
                flask.jsonify(message="cannot find registry for repository %s" % repository), 400))
        content['docker']['registry'] = registry
    elif content['docker']['registry'] not in service_module._config['docker']['registries']:
        flask.abort(flask.make_response(flask.jsonify(message="unknown docker registry"), 400))
    resource = service_module.get_resource_from_options(content["options"])

    iterations = 1
    if "iterations" in content:
        iterations = content["iterations"]
        if (task_type != "train" and iterations != 1) or iterations < 1:
            flask.abort(flask.make_response(flask.jsonify(message="invalid value for iterations"), 400))

    priority = content.get("priority", 0)

    (xxyy, parent_task_id) = shallow_command_analysis(content["docker"]["command"])

    task_ids = []

    while iterations > 0:
        task_id = build_task_id(content, xxyy, parent_task_id)
        task.create(redis, task_id, task_type, parent_task_id, resource, service, content, files, priority)
        task_ids.append(task_id)
        iterations -= 1
        if iterations > 0:
            parent_task_id = task_id
            change_parent_task(content["docker"]["command"], parent_task_id)

    if len(task_ids) == 1:
        task_ids = task_ids[0]

    return flask.jsonify(task_ids)

@task_request
@app.route("/task/status/<string:task_id>", methods=["GET"])
def status(task_id):
    response = task.info(redis, task_id, [])
    return flask.jsonify(response)

@task_request
@app.route("/task/<string:task_id>", methods=["DELETE"])
def del_task(task_id):
    response = task.delete(redis, task_id)
    if isinstance(response, list) and not response[0]:
        flask.abort(flask.make_response(flask.jsonify(message=response[1]), 400))        
    return flask.jsonify(message="deleted %s" % task_id)

@app.route("/task/list/<string:pattern>", methods=["GET"])
def list_tasks(pattern):
    ltask = []
    for task_key in task.scan_iter(redis, pattern):
        task_id = task.id(task_key)
        info = task.info(redis, task_id,
                ["queued_time", "resource", "content", "status", "message", "type", "iterations", "priority"])
        if info["content"] is not None and info["content"] != "":
            content = json.loads(info["content"])
            info["image"] = content['docker']['image']
            del info['content']
        else:
            info["image"] = '-'
        info['task_id'] = task_id
        ltask.append(info)
    return flask.jsonify(ltask)

@task_request
@app.route("/task/terminate/<string:task_id>", methods=["GET"])
def terminate(task_id):
    with redis.acquire_lock(task_id):
        current_status = task.info(redis, task_id, "status")
        if current_status is None:
            flask.abort(flask.make_response(flask.jsonify(message="task %s unknown" % task_id), 404))
        elif current_status == "stopped":
            return flask.jsonify(message="%s already stopped" % task_id)
        phase = flask.request.args.get('phase')
        task.terminate(redis, task_id, phase=phase)
    return flask.jsonify(message="terminating %s" % task_id)

@task_request
@app.route("/task/beat/<string:task_id>", methods=["PUT"])
def task_beat(task_id):
    duration = flask.request.args.get('duration')
    try:
        if duration is not None:
            duration = int(duration)
    except ValueError:
        flask.abort(flask.make_response(flask.jsonify(message="invalid duration value"), 400))
    container_id = flask.request.args.get('container_id')
    task.beat(redis, task_id, duration, container_id)
    return flask.jsonify(200)

@task_request
@app.route("/beat/<string:task_id>", methods=["GET"])
def task_beat_old(task_id):
    task_beat(task_id)

@task_request
@app.route("/task/file/<string:task_id>/<string:filename>", methods=["GET"])
def get_file(task_id, filename):
    content = task.get_file(redis, task_id, filename)
    if content is None:
        flask.abort(flask.make_response(
            flask.jsonify(message="cannot find file %s for task %s" % (filename, task_id)), 404))
    response = flask.make_response(content)
    return response

@task_request
@app.route("/task/file/<string:task_id>/<string:filename>", methods=["POST"])
def post_file(task_id, filename):
    content = flask.request.get_data()
    task.set_file(redis, task_id, content, filename)
    return flask.jsonify(200)

@task_request
@app.route("/task/log/<string:task_id>", methods=["GET"])
def get_log(task_id):
    content = task.get_log(redis, task_id)
    if content is None:
        flask.abort(flask.make_response(
            flask.jsonify(message="cannot find log for task %s" % task_id), 404))
    response = flask.make_response(content)
    return response

@task_request
@app.route("/task/log/<string:task_id>", methods=["PATCH"])
def append_log(task_id):
    content = flask.request.get_data()
    task.append_log(redis, task_id, content)
    return flask.jsonify(200)

@task_request
@app.route("/task/log/<string:task_id>", methods=["POST"])
def post_log(task_id):
    content = flask.request.get_data()
    task.set_log(redis, task_id, content)
    return flask.jsonify(200)
